crates/e2store/src/e2hs.rs

```diff
@@ -0,0 +1,339 @@
+                "invalid block index entry: incorrect header length: found {}, expected {}",
+                entry.header.length, 65552
+            )
+        );
+        ensure!(
+            entry.header.reserved == 0,
+            "invalid block index entry: incorrect header reserved bytes"
+        );
+        ensure!(
+            entry.value.len() == 65552,
+            "invalid block index entry: incorrect value length"
+        );
+        Ok(Self {
+            block_index: BlockIndex::try_from(entry.clone())?,
+        })
+    }
+}
+
+impl TryInto<Entry> for E2HSBlockIndexEntry {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> Result<Entry, Self::Error> {
+        let mut buf: Vec<u64> = vec![];
+        buf.push(self.block_index.starting_number);
+        buf.extend_from_slice(&self.block_index.indices);
+        buf.push(self.block_index.count);
+        let encoded = buf
+            .iter()
+            .flat_map(|i| i.to_le_bytes().to_vec())
+            .collect::<Vec<u8>>();
+        Ok(Entry::new(entry_types::BLOCK_INDEX, encoded))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::e2store::types::Header;
+
+    #[test]
+    fn test_e2hs_round_trip() {
+        let raw_e2hs = fs::read("../../test_assets/era1/mainnet-00000-d4e56740.e2hs").unwrap();
+        let e2hs = E2HS::deserialize(&raw_e2hs).expect("failed to deserialize e2hs");
+        let raw_e2hs2 = e2hs.write().expect("failed to serialize e2hs");
+        assert_eq!(raw_e2hs, raw_e2hs2);
+    }
+
+    #[rstest::rstest]
+    #[case(0)]
+    #[case(1)]
+    #[case(100)]
+    #[case(8191)]
+    fn test_e2hs_block_index(#[case] block_number: u64) {
+        let raw_e2hs = fs::read("../../test_assets/era1/mainnet-00000-d4e56740.e2hs").unwrap();
+        let block_tuple = E2HS::get_tuple_by_index(&raw_e2hs, block_number).unwrap();
+        assert_eq!(
+            block_tuple
+                .header_with_proof
+                .header_with_proof
+                .header
+                .number,
+            block_number
+        );
+    }
+
+    #[rstest::rstest]
+    #[case(0)]
+    #[case(1)]
+    #[case(100)]
+    #[case(8191)]
+    fn test_e2hs_block_index_direct_access(#[case] block_number: u64) {
+        let raw_e2hs = fs::read("../../test_assets/era1/mainnet-00000-d4e56740.e2hs").unwrap();
+        let file = E2StoreMemory::deserialize(&raw_e2hs).expect("invalid e2hs file");
+        let block_index =
+            E2HSBlockIndexEntry::try_from(file.entries.last().expect("missing block index entry"))
+                .expect("invalid block index entry")
+                .block_index;
+        let index = block_index.indices[block_number as usize];
+        let header_bytes = raw_e2hs
+            .get((index as usize)..(index as usize + 8))
+            .unwrap();
+        let header = Header::deserialize(header_bytes).expect("invalid header");
+        let hwp_length = header.length as u64;
+        let hwp_bytes = raw_e2hs
```

Comment:
I am confused I don't think these are headers with proof no?

Reply:
.... i'm pretty sure they are? i deserialize it just a few lines below

Reply:
Oh my bad I confusing the imports `use crate::e2store::types::Header;` and `use crate::types::Header;`